sql/permissions.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/structured"
+)
+
+// Permissions describes the full set of permissions for a descriptor.
+// This allows quick lookup for a given privilege and user.
+type Permissions struct {
+	read  map[string]struct{}
+	write map[string]struct{}
+}
```

Comment:
I think the common use case for permissions is point lookups as to whether a particular user has a particular permission. In that case, using `sort.SearchStrings` will be faster than creating these maps. And it also might be more natural to add `structured.DatabaseDescriptor.{Grant,Revoke}` methods. These could use `sort.SearchStrings` to insert into the sorted slice of users.